### PR TITLE
HIVE-21375 initial commit

### PR DESCRIPTION
**Pull request by szlta**

### CODE REVIEW

File: hcatalog/streaming/src/java/org/apache/hive/hcatalog/streaming/HiveEndPoint.java [hunk 1]
```diff
@@ -394,10 +404,12 @@ public Void run() throws Exception {
                 return null;
               }
             });
-        try {
-          FileSystem.closeAllForUGI(ugi);
-        } catch (IOException exception) {
-          LOG.error("Could not clean up file-system handles for UGI: " + ugi, exception);
+        if (decRefForUgi(ugi) == 0) {
+          try {
+              FileSystem.closeAllForUGI(ugi);
```
Comment (pvary) [inline hcatalog/streaming/src/java/org/apache/hive/hcatalog/streaming/HiveEndPoint.java:409]: nit: indent

File: hcatalog/streaming/src/java/org/apache/hive/hcatalog/streaming/HiveEndPoint.java [hunk 2]
```diff
@@ -1107,4 +1114,37 @@ private static void setHiveConf(HiveConf conf, HiveConf.ConfVars var, boolean va
     conf.setBoolVar(var, value);
   }
 
+  private static void incRefForUgi(UserGroupInformation ugi) {
+    refCountLock.lock();
+    try {
+      Long prevCount = ugiConnectionRefCount.putIfAbsent(ugi, 1L);
+      if (prevCount != null) {
```
Comment (pvary) [inline hcatalog/streaming/src/java/org/apache/hive/hcatalog/streaming/HiveEndPoint.java:1121]: How can this be null?

Comment (pvary) [reply hcatalog/streaming/src/java/org/apache/hive/hcatalog/streaming/HiveEndPoint.java:1121]: Sorry my mistake

**Review by pvary — APPROVED** [review]

**Comment by szlta** [comment]
Cherry-picked changes here that allow test run on this PR. There were many test failures as branch-3 is in bad shape currently, but none of those seemed to be related to hcatalog streaming, thus I'm moving forward with this change.

**Pull request merged**